compiler/rustc_typeck/src/check/method/mod.rs

```diff
@@ -511,6 +516,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         };
         obligations.extend(traits::predicates_for_generics(cause.clone(), self.param_env, bounds));
 
+        // If we're producing obligations for an operator trait, and it has an `Output`
+        // associated type, then we add an `Output=expected` obligation.
+        if let (true, Expectation::ExpectHasType(output_ty)) = (is_op, expected) {
```

Comment:
nits: `let` chains can make this more readable, and `Expectation::only_has_type` might be useful instead of matching on the variant itself.